streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java

```diff
@@ -0,0 +1,1123 @@
+                        }
+
+                        return new KeyValue<>(key, state.get(key));
+                    }
+
+                    @Override
+                    public void close() { }
+                };
+            } }, storeNames)
+            .to(MULTI_PARTITION_OUTPUT_TOPIC);
+
+        final Properties properties = new Properties();
+        properties.put(StreamsConfig.CLIENT_ID_CONFIG, appDir);
+        properties.put(StreamsConfig.PROCESSING_GUARANTEE_CONFIG, processingGuarantee);
+        properties.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, Long.MAX_VALUE);
+        properties.put(StreamsConfig.consumerPrefix(ConsumerConfig.METADATA_MAX_AGE_CONFIG), "1000");
+        properties.put(StreamsConfig.consumerPrefix(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG), "earliest");
+        properties.put(StreamsConfig.consumerPrefix(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG), 5 * 1000);
+        properties.put(StreamsConfig.consumerPrefix(ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG), 5 * 1000 - 1);
+        properties.put(StreamsConfig.consumerPrefix(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG), MAX_POLL_INTERVAL_MS);
+        // TODO
+        //   if we don't use this custom partitioner the test fails for the non-error case
+        //   unclear why -- see other TODO
+        properties.put(StreamsConfig.producerPrefix(ProducerConfig.PARTITIONER_CLASS_CONFIG), KeyPartitioner.class);
+        properties.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
+        properties.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath() + File.separator + appDir);
+
+        final Properties config = StreamsTestUtils.getStreamsConfig(
+            applicationId,
+            CLUSTER.bootstrapServers(),
+            Serdes.LongSerde.class.getName(),
+            Serdes.LongSerde.class.getName(),
+            properties
+        );
+
+        final KafkaStreams streams = new KafkaStreams(builder.build(), config, new TestKafkaClientSupplier());
+
+        streams.setUncaughtExceptionHandler((t, e) -> {
+            if (uncaughtException != null) {
+                e.printStackTrace(System.err);
+                fail("Should only get one uncaught exception from Streams.");
+            }
+            uncaughtException = e;
+        });
+
+        return streams;
+    }
+
+    private void waitForStateTransition(final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> observed,
+                                        final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> expected)
+        throws Exception {
+
+        try {
+            waitForCondition(
+                () -> observed.equals(expected),
+                MAX_WAIT_TIME_MS,
+                "Client did not startup on time."
+            );
+        } catch (final AssertionError error) {
+            final AssertionError newError = new AssertionError("Client transitions: " + observed);
```

Comment:
To get the correct `observed` state transition, we need to add them _after_ `waitForCondition` failed... Originally, we just called:
```
waitForCondition(
                () -> observed.equals(expected),
                MAX_WAIT_TIME_MS,
                "Client did not startup on time." + observed
            );
```
but this creates the error string with `observed` when started and thus the error message is miss-leading and useless